# pandas

## Getting Started


In [1]:
#Importing the package
import pandas as pd

## Series

* One dimensional array with axis labels
* Can store any data type


### Creating Series

In [2]:
s = pd.Series()
print s

Series([], dtype: float64)


### Making Series Out of Lists

In [3]:
s = pd.Series([109,289,982])
print s

0    109
1    289
2    982
dtype: int64


### Changing the Index Names

In [5]:
s = pd.Series([109,289,982],index=["a","b","c"])
print s

a    109
b    289
c    982
dtype: int64


## Making Series Out of Dictionaries

In [8]:
data = {'a' : 0., 'b' : 1., 'c' : 2.}
s = pd.Series(data)
print s

a    0.0
b    1.0
c    2.0
dtype: float64


In [9]:
s = pd.Series(data,index=['b','c','d','a'])
print s

b    1.0
c    2.0
d    NaN
a    0.0
dtype: float64


# Data Frames

## Syntax

In [11]:
df = pd.DataFrame()
print df

Empty DataFrame
Columns: []
Index: []


## One Dimensional

In [12]:
data = [1,2,3,4,5]
df = pd.DataFrame(data)
print df

   0
0  1
1  2
2  3
3  4
4  5


## Two Dimensional

### By List

In [13]:
data = [['Alex',10],['Bob',12],['Clarke',13]]
df = pd.DataFrame(data,columns=['Name','Age'])
print df

     Name  Age
0    Alex   10
1     Bob   12
2  Clarke   13


### By Dictionary

In [14]:
data = {'Name':['Tom', 'Jack', 'Steve', 'Ricky'],'Age':[28,34,29,42]}
df = pd.DataFrame(data)
print df

   Age   Name
0   28    Tom
1   34   Jack
2   29  Steve
3   42  Ricky


### Changing Index

In [15]:
data = {'Name':['Tom', 'Jack', 'Steve', 'Ricky'],'Age':[28,34,29,42]}
df = pd.DataFrame(data, index=['rank1','rank2','rank3','rank4'])
print df

       Age   Name
rank1   28    Tom
rank2   34   Jack
rank3   29  Steve
rank4   42  Ricky


## Creating Columns

In [16]:
data = [{'a': 1, 'b': 2},{'a': 5, 'b': 10, 'c': 20}]
df = pd.DataFrame(data)
print df

   a   b     c
0  1   2   NaN
1  5  10  20.0


In [17]:
data = [{'a': 1, 'b': 2},{'a': 5, 'b': 10, 'c': 20}]

#With two column indices, values same as dictionary keys
df1 = pd.DataFrame(data, index=['first', 'second'], columns=['a', 'b'])


In [18]:
#With two column indices with one index with other name
df2 = pd.DataFrame(data, index=['first', 'second'], columns=['a', 'b1'])
print df1
print df2

        a   b
first   1   2
second  5  10
        a  b1
first   1 NaN
second  5 NaN


## Indexing Columns

In [19]:
d = {'one' : pd.Series([1, 2, 3], index=['a', 'b', 'c']),
      'two' : pd.Series([1, 2, 3, 4], index=['a', 'b', 'c', 'd'])}

df = pd.DataFrame(d)
print df

   one  two
a  1.0    1
b  2.0    2
c  3.0    3
d  NaN    4


In [20]:
print df ['one']

a    1.0
b    2.0
c    3.0
d    NaN
Name: one, dtype: float64


## More Column Fun

In [21]:
# Adding a new column to an existing DataFrame object with column label by passing new series

print ("Adding a new column by passing as Series:")
df['three']=pd.Series([10,20,30],index=['a','b','c'])
print df

print ("Adding a new column using the existing columns in DataFrame:")
df['four']=df['one']+df['three']

print df

Adding a new column by passing as Series:
   one  two  three
a  1.0    1   10.0
b  2.0    2   20.0
c  3.0    3   30.0
d  NaN    4    NaN
Adding a new column using the existing columns in DataFrame:
   one  two  three  four
a  1.0    1   10.0  11.0
b  2.0    2   20.0  22.0
c  3.0    3   30.0  33.0
d  NaN    4    NaN   NaN


In [22]:
print ("Our dataframe is:")
print df

# using del function
print ("Deleting the first column using DEL function:")
del df['one']
print df


Our dataframe is:
   one  two  three  four
a  1.0    1   10.0  11.0
b  2.0    2   20.0  22.0
c  3.0    3   30.0  33.0
d  NaN    4    NaN   NaN
Deleting the first column using DEL function:
   two  three  four
a    1   10.0  11.0
b    2   20.0  22.0
c    3   30.0  33.0
d    4    NaN   NaN


In [23]:
# using pop function
print ("Deleting another column using POP function:")
df.pop('two')
print df

Deleting another column using POP function:
   three  four
a   10.0  11.0
b   20.0  22.0
c   30.0  33.0
d    NaN   NaN


## Indexing Rows

### By Name

In [24]:
d = {'one' : pd.Series([1, 2, 3], index=['a', 'b', 'c']), 
     'two' : pd.Series([1, 2, 3, 4], index=['a', 'b', 'c', 'd'])}

df = pd.DataFrame(d)
print df.loc['b']

one    2.0
two    2.0
Name: b, dtype: float64


### By Index

In [25]:
print df.iloc[2]

one    3.0
two    3.0
Name: c, dtype: float64


### Splicing

In [26]:
df = pd.DataFrame(d)
print df[2:4]

   one  two
c  3.0    3
d  NaN    4


### More Row Fun!

In [27]:
df = pd.DataFrame([[1, 2], [3, 4]], columns = ['a','b'])
df2 = pd.DataFrame([[5, 6], [7, 8]], columns = ['a','b'])

df = df.append(df2)
print df

   a  b
0  1  2
1  3  4
0  5  6
1  7  8


In [28]:
df = df.drop(0)
print df

   a  b
1  3  4
1  7  8


## CSV Work!

In [29]:
#Read the file
data = pd.read_csv("chopstick-effectiveness.csv")

#Convert all the strings to numbers
for col in data.columns:
    data[col] = pd.to_numeric(data[col])

In [41]:
data

Food Pinching Efficiency  Individual  Chopstick Length
0                       19.55           1               180
1                       27.24           2               180
2                       28.76           3               180
3                       31.19           4               180
4                       21.91           5               180
5                       27.62           6               180
6                       29.46           7               180
7                       26.35           8               180
8                       26.69           9               180
9                       30.22          10               180
10                      27.81          11               180
11                      23.46          12               180
12                      23.64          13               180
13                      27.85          14               180
14                      20.62          15               180
15                      25.35          16               180
16                      28.00          17               180
17                      23.49          18               180
18                      27.77          19               180
19                      18.48          20               180
20                      23.01          21               180
21                      22.66          22               180
22                      23.24          23               180
23                      22.82          24               180
24                      17.94          25               180
25                      26.67          26               180
26                      28.98          27               180
27                      21.48          28               180
28                      14.47          29               180
29                      28.29          30               180
..                        ...         ...               ...
156                     26.18           2               330
157                     25.93           3               330
158                     28.61           4               330
159                     20.54           5               330
160                     26.44           6               330
161                     29.36           7               330
162                     19.77           8               330
163                     31.69           9               330
164                     24.64          10               330
165                     22.09          11               330
166                     23.42          12               330
167                     28.63          13               330
168                     26.30          14               330
169                     22.89          15               330
170                     22.68          16               330
171                     30.92          17               330
172                     20.74          18               330
173                     27.24          19               330
174                     17.12          20               330
175                     23.63          21               330
176                     20.91          22               330
177                     23.49          23               330
178                     24.86          24               330
179                     16.28          25               330
180                     21.52          26               330
181                     27.22          27               330
182                     17.41          28               330
183                     16.42          29               330
184                     28.22          30               330
185                     27.52          31               330

[186 rows x 3 columns]

In [42]:
#Shows part of the data
data.head()

Food Pinching Efficiency  Individual  Chopstick Length
0                     19.55           1               180
1                     27.24           2               180
2                     28.76           3               180
3                     31.19           4               180
4                     21.91           5               180

## Cleaning/Organizing Data

In [43]:
#Cleaning out bad rows by adding dummy values
data = data.fillna(0) 

In [44]:
#Cleaning out bad rows by dropping the row
data = data.dropna() 

In [45]:
#Picking out certain data
data[data['Food Pinching Efficiency']>20]

Food Pinching Efficiency  Individual  Chopstick Length
1                       27.24           2               180
2                       28.76           3               180
3                       31.19           4               180
4                       21.91           5               180
5                       27.62           6               180
6                       29.46           7               180
7                       26.35           8               180
8                       26.69           9               180
9                       30.22          10               180
10                      27.81          11               180
11                      23.46          12               180
12                      23.64          13               180
13                      27.85          14               180
14                      20.62          15               180
15                      25.35          16               180
16                      28.00          17               180
17                      23.49          18               180
18                      27.77          19               180
20                      23.01          21               180
21                      22.66          22               180
22                      23.24          23               180
23                      22.82          24               180
25                      26.67          26               180
26                      28.98          27               180
27                      21.48          28               180
29                      28.29          30               180
30                      27.97          31               180
31                      23.53           1               210
32                      26.39           2               210
33                      30.90           3               210
..                        ...         ...               ...
149                     22.86          26               300
150                     25.09          27               300
153                     30.91          30               300
154                     25.92          31               300
155                     21.32           1               330
156                     26.18           2               330
157                     25.93           3               330
158                     28.61           4               330
159                     20.54           5               330
160                     26.44           6               330
161                     29.36           7               330
163                     31.69           9               330
164                     24.64          10               330
165                     22.09          11               330
166                     23.42          12               330
167                     28.63          13               330
168                     26.30          14               330
169                     22.89          15               330
170                     22.68          16               330
171                     30.92          17               330
172                     20.74          18               330
173                     27.24          19               330
175                     23.63          21               330
176                     20.91          22               330
177                     23.49          23               330
178                     24.86          24               330
180                     21.52          26               330
181                     27.22          27               330
184                     28.22          30               330
185                     27.52          31               330

[165 rows x 3 columns]

In [46]:
#Sorting by chopstick length
data.sort_values("Chopstick Length")

Food Pinching Efficiency  Individual  Chopstick Length
0                       19.55           1               180
30                      27.97          31               180
29                      28.29          30               180
28                      14.47          29               180
27                      21.48          28               180
26                      28.98          27               180
25                      26.67          26               180
24                      17.94          25               180
23                      22.82          24               180
21                      22.66          22               180
20                      23.01          21               180
19                      18.48          20               180
18                      27.77          19               180
17                      23.49          18               180
16                      28.00          17               180
22                      23.24          23               180
14                      20.62          15               180
15                      25.35          16               180
2                       28.76           3               180
3                       31.19           4               180
4                       21.91           5               180
5                       27.62           6               180
6                       29.46           7               180
7                       26.35           8               180
1                       27.24           2               180
9                       30.22          10               180
10                      27.81          11               180
11                      23.46          12               180
12                      23.64          13               180
13                      27.85          14               180
..                        ...         ...               ...
173                     27.24          19               330
174                     17.12          20               330
175                     23.63          21               330
176                     20.91          22               330
183                     16.42          29               330
178                     24.86          24               330
179                     16.28          25               330
180                     21.52          26               330
181                     27.22          27               330
182                     17.41          28               330
171                     30.92          17               330
177                     23.49          23               330
170                     22.68          16               330
162                     19.77           8               330
168                     26.30          14               330
167                     28.63          13               330
166                     23.42          12               330
165                     22.09          11               330
164                     24.64          10               330
163                     31.69           9               330
161                     29.36           7               330
160                     26.44           6               330
159                     20.54           5               330
158                     28.61           4               330
157                     25.93           3               330
156                     26.18           2               330
155                     21.32           1               330
184                     28.22          30               330
169                     22.89          15               330
185                     27.52          31               330

[186 rows x 3 columns]

In [47]:
#Sorting by the individual
data.sort_values("Individual").head()

Food Pinching Efficiency  Individual  Chopstick Length
0                       19.55           1               180
155                     21.32           1               330
124                     22.50           1               300
31                      23.53           1               210
93                      24.40           1               270

## Running Analyses

In [48]:
data.describe()

Food Pinching Efficiency  Individual  Chopstick Length
count                186.000000  186.000000         186.00000
mean                  25.005591   16.000000         255.00000
std                    4.039693    8.968413          51.37304
min                   14.470000    1.000000         180.00000
25%                   22.540000    8.000000         210.00000
50%                   24.905000   16.000000         255.00000
75%                   27.932500   24.000000         300.00000
max                   36.150000   31.000000         330.00000

In [49]:
data.sum()

Food Pinching Efficiency     4651.04
Individual                   2976.00
Chopstick Length            47430.00
dtype: float64

In [50]:
data.std()

Food Pinching Efficiency     4.039693
Individual                   8.968413
Chopstick Length            51.373040
dtype: float64

In [51]:
data.mean()

Food Pinching Efficiency     25.005591
Individual                   16.000000
Chopstick Length            255.000000
dtype: float64

## So which length of chopsticks is most efficient?

In [52]:
#Get a list of unique lengths of chopsticks
lengths = data["Chopstick Length"].unique()

#Loop through and get the average efficiencies
for length in lengths:
    df = data[data["Chopstick Length"] == length]
    print length
    print df["Food Pinching Efficiency"].mean()
    print "==================\n"

180
24.9351612903

210
25.4838709677

240
26.3229032258

270
24.3238709677

300
24.9680645161

330
23.9996774194



### We can see that the chopsticks of length 240 is most efficient!

## But who was the better user?

In [53]:
#Get a list of unique individuals
peeps = data["Individual"].unique()
efficiencies = []

#Loop through and get the average efficiencies
for person in peeps:
    df = data[data["Individual"] == person]
    print person
    avg = df["Food Pinching Efficiency"].mean()
    efficiencies.append(avg)
    print avg
    print "==================\n"

1
22.1066666667

2
26.455

3
29.1283333333

4
27.5566666667

5
21.235

6
28.0683333333

7
30.005

8
21.0083333333

9
31.0283333333

10
28.0683333333

11
24.9716666667

12
24.915

13
26.92

14
27.4033333333

15
22.9466666667

16
24.6583333333

17
27.4933333333

18
23.2216666667

19
29.265

20
20.0966666667

21
23.2233333333

22
22.4383333333

23
23.5733333333

24
25.0866666667

25
19.0416666667

26
23.3266666667

27
27.2766666667

28
19.835

29
17.4716666667

30
30.12

31
27.2283333333



In [54]:
print efficiencies.index(max(efficiencies)) + 1

9


## Person 9 can use chopsticks the best!